src/map/script.c

```diff
@@ -18318,33 +18318,66 @@ static BUILDIN(swap)
 
 // [zBuffer] List of dynamic var commands --->
 
+/**
+ * Sets the value of a variable or an array.
+ * The variable's name can be passed as a string and thus can dynamically be constructed.
+ *
+ * [Kenpachi]
+ * Passing a reference to a variable is supported, too.
+ * Useful when using getvariableofnpc(), getvariableofpc(), getd() or getarg().
+ *
+ * @code{.herc}
+ *	setd("<variable name>", <value>);
+ *	setd(<reference to variable>, <value>);
+ * @endcode
+ *
+ */
 static BUILDIN(setd)
 {
-	struct map_session_data *sd = NULL;
+	int idx = 0;
+	int length = 0;
 	char varname[100];
 	const char *buffer;
-	int elem;
-	buffer = script_getstr(st, 2);
+	const void *value;
+	struct map_session_data *sd = NULL;
+	struct script_data *data = script_getdata(st, 2);
+	struct reg_db *ref = reference_getref(data);
 
-	if(sscanf(buffer, "%99[^[][%d]", varname, &elem) < 2)
-		elem = 0;
+	if (data_isreference(data)) {
+		buffer = reference_getname(data);
+		idx = reference_getindex(data);
 
-	if( not_server_variable(*varname) )
-	{
-		sd = script->rid2sd(st);
-		if( sd == NULL )
-		{
-			ShowError("script:setd: no player attached for player variable '%s'\n", buffer);
-			return true;
-		}
+		if (not_server_variable(*buffer))
+			ref = NULL;
+	} else {
+		buffer = script_getstr(st, 2);
 	}
 
-	if (is_string_variable(varname)) {
-		script->setd_sub(st, sd, varname, elem, script_getstr(st, 3), NULL);
-	} else {
-		script->setd_sub(st, sd, varname, elem, (const void *)h64BPTRSIZE(script_getnum(st, 3)), NULL);
+	/// Special case: Variable of type string was passed directly or by getd().
+	if (data_isreference(data) && reference_tovariable(data) && script_isstringtype(st, 2)) {
+		if (data->type == C_STR) /// Type has changed now, bacause script->get_val() was called by script_isstringtype().
+			buffer = data->u.str; /// Get the value of the passed variable instead of its name.
 	}
 
+	if (sscanf(buffer, "%99[^[][%d]", varname, &idx) < 2 && idx == 0)
```

Comment:
even this is old code, probably better add size limit for ``%d`` because it can be too big and overflow
for example replace to this ``%30d``

Reply:
Added and tested.